cmd/jujuc/main.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2012-2014 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/exec"
+	"gopkg.in/juju/names.v3"
+
+	"github.com/juju/juju/juju/sockets"
+)
+
+var logger = loggo.GetLogger("juju.cmd.jujud")
+
+func init() {
+	rand.Seed(time.Now().UTC().UnixNano())
+}
+
+// TODO: This becomes jujucDoc describing the hook tools
+var jujucDoc = `
+The jujuc command forwards invocations over RPC for execution by the
+juju unit agent. When used in this way, it expects to be called via a symlink
+named for the desired remote command, and expects JUJU_AGENT_SOCKET_ADDRESS and
+JUJU_CONTEXT_ID be set in its model.
+`
+
+const (
+	// exit_err is the value that is returned when the user has run juju in an invalid way.
+	exit_err = 2
+	// exit_panic is the value that is returned when we exit due to an unhandled panic.
+	exit_panic = 3
+)
+
+func getenv(name string) (string, error) {
+	value := os.Getenv(name)
+	if value == "" {
+		return "", errors.Errorf("%s not set", name)
+	}
+	return value, nil
+}
+
+func getwd() (string, error) {
+	dir, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+	abs, err := filepath.Abs(dir)
+	if err != nil {
+		return "", err
+	}
+	return abs, nil
+}
+
+func getSocket() (sockets.Socket, error) {
```

Comment:
perhaps this function can be shared between jujuc and jujud as it is non trivial and having a common implementation would be good, maybe a helper in github.com/juju/juju/juju/sockets

Reply:
indeed. we'll want to be a little careful about dependencies, but somewhere like 'sockets' seems a reasonable place (and we already depend on it).

It also feels like something I'd really like to decouple from env vars for testability, and then the values passed in could come from env vars.